Support selective querying of stats for `getblockstats` + fix overflows

zzantares opened the pull request:
This basically allows downstream users to specify only the statistics they
need from the `getblockstats` response, previous to this the caller got every
field but the RPC command allows selective stat selection and I feel the
client we expose should too.

Another fix packed in this PR is that when querying stats for blocks with big
fees I observed errors like:

```
DecodingError "Error in $.totalfee: parsing Word32 failed, value is either floating or will cause over or underflow 8.521634113e9"
```

Same for `maxfee`, so we need to use wider types for these fields. It is worth
mentioning that in bitcoin-core these are `int64_t` however I'm not sure we
blindly want to match these types since it would be very weird to observe
negative values on these fields.

zzantares committed:
Allow optional stats in `getblockstats` response

This basically allows downstream users to specify only the statistics they
need from the `getblockstats` response, previous to this the caller got every
field but the RPC command allows selective stat selection and I feel we should too.

zzantares committed:
Wider types for `totalfee` and `maxfee` block stats

When querying stats for some big blocks I observed errors like:

```
DecodingError "Error in $.totalfee: parsing Word32 failed, value is either floating or will cause over or underflow 8.521634113e9"
```

Same for `maxfee`, so we need to use wider types for these fields. It is worth
mentioning that in bitcoin-core these are `int64_t` however I'm not sure we
blindly want to match these types since it would be very weird to observe
negative values on these fields.

GambolingPangolin reviewed (commented):
I would really prefer not to have ticked versions of bindings.  We can bump the version and export `getBlockStats` with signature
```
getBlockStats :: BlockHash -> Maybe [Text] -> BitcoindClient BlockStats
```

After all, this is already a breaking change because `BlockStats` gets a different type.

zzantares commented:
@GambolingPangolin I made the change, however now I'm thinking if this change should really bump the version? since we actually haven't made an official release of `0.3.0.0`.

GambolingPangolin commented:
> @GambolingPangolin I made the change, however now I'm thinking if this change should really bump the version? since we actually haven't made an official release of `0.3.0.0`.

In that case we can leave it at 3.0.0.0

wraithm commented on bitcoind-rpc/src/Bitcoin/Core/RPC/Blockchain.hs:
```diff
@@ -336,7 +337,11 @@ getBlockFilter :: BlockHash -> BitcoindClient CompactFilter
+{- | Compute per block statistics for a given window. All amounts are in
+ satoshis.  It won't work for some heights with pruning.
+-}
+getBlockStats :: BlockHash -> Maybe [Text] -> BitcoindClient BlockStats
```
Should we have an enum for the field names?

GambolingPangolin replied:
That's not a bad idea.  Is it stable across recent versions?

zzantares replied:
I'll check against our supported versions 👍🏼

zzantares replied:
I've made the change and tested against the versions we support, there's something fishy against `bitcoin-core-0.19.1` the tests never complete but I confirmed this is unrelated to these changes as the same happens on `master`, all other versions work just fine, I'll take a deeper look into why that happens.

zzantares committed:
Make `getBlockStats` to allow selective stats by default

Instead of using `getBlockStats'` for selective queries and `getBlockStats`
for non-selective queries, we want `getBlockStats` to be the only one version
of this function, since this is now considered a breaking change I bumped the
version of the library.

zzantares committed:
Use ADTs for specifying statistics in `getblockstats` RPC command

Instead of letting the user to specify any text, we explicitly list the stats
supported by our library as an ADT.

zzantares committed:
Add `Stat` prefix to `BlockStat` data constructors

This reduces the possibility of name clashing and makes it more clear when
users do qualified imports, e.g. `RPC.StatSubsidy` tells a bit more than just
`RPC.Subsidy`.

zzantares commented:
Added a new commit that just prefixes data constructors with `Stat`, I noticed when using this that `RPC.StatTxs` or `RPC.StatSubsidy` conveys a bit more meaning than `RPC.Txs` or `RPC.Subsidy`.

GambolingPangolin commented:
> Added a new commit that just prefixes data constructors with `Stat`, I noticed when using this that `RPC.StatTxs` or `RPC.StatSubsidy` conveys a bit more meaning than `RPC.Txs` or `RPC.Subsidy`.

This is a really nice change.  Thanks!
